README.md
# Daioikachan

[Ikachan](https://github.com/yappo/p5-App-Ikachan) compatible interface with multiple backends (IRC, Slack, etc).

## Requirements

* Ruby

## Installation

Write Gemfile:

```
source "https://rubygems.org"

gem 'daioikachan'
```

Run:

```
$ bundle
```

IRC, Slack backends are bundled as default.

Daioikachan supports plugin architecture. You may add your favirite backends as gems. See Plugin section for details.

## How to Run

This is an example to post messages for both IRC and Slack.

### Start

Generate a sample [daioikachan.conf](./examples/example.conf):

```
$ bundle exec daioikachan -g daioikachan.conf
```

Create `.env` and configure your IRC server and Slack token:

```
IRC_SERVER=XX.XX.XX.XX
SLACK_API_TOKEN=XXX-XXXXX-XXXXXX-XXXXX
````

Start `daioikachan` as

```
$ bundle exec daioikachan
```

### Test

Test to post a message to `#daioikachan` channel of both IRC and Slack via `daioikachan` like:

```
$ curl -d "channel=#daioikachan&message=test daioikachan" http://localhost:4979/notice
```

Look whether posting to IRC, and Slack succeeds.

## Configuration

See [example.conf](./example.conf) or [multi_slack.conf](./examples/multi_slack.conf) as examples.

You might've noticed that the config file is similar with Fluentd.
Yes, `daioikachan` is created based on Fluentd.
So, you can use `routing` functions which Fluentd has (`tag` and `label`).

See [config-file](http://docs.fluentd.org/articles/config-file) documentation of Fluentd for details.

See following pages for built-in plugins.

* [in_daioikachan](./README/in_daioikachan.md)
* [fluent-plugin-irc](https://github.com/choplin/fluent-plugin-irc)
* [fluent-plugin-slack](https://github.com/sowawa/fluent-plugin-slack)

If you need other backends, search other fluentd output plugins and add them. Enjoy!

## Plugin

You can create and add your own backends for daioikachan as a Fluentd Plugin.

See http://www.fluentd.org/plugins#notifications for available plugins.

## API

### /notice

Send `notice` message.

```
$ curl -d "channel=#channel&message=test message" http://localhost:4979/notice
```

### /privmsg

Send `privmsg` message.

```
$ curl -d "channel=#channel&message=test message" http://localhost:4979/privmsg
```

### /join

The server always returns 200, and ignore.

IRC client (`fluent-plugin-irc`) automatically joins to a channel on sending message.
Slack client does not require to join to a channel.

### /leave

The server always returns 200, and ignores.

## ChangeLog

See [CHANGELOG.md](CHANGELOG.md) for details.

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new [Pull Request](../../pull/new/master)

## Copyright

### daioikachan, and fluent-plugin-daioikachan

Copyright (c) 2015 Naotoshi Seo. See [LICENSE](LICENSE) for details.

### fluent-plugin-irc

See https://github.com/choplin/fluent-plugin-irc

### fluent-plugin-slack

See https://github.com/sowawa/fluent-plugin-slack


CHANGELOG.md
# 0.0.9 (2015/10/15)

Fixes

* require `rack` (thanks to maroekun)

# 0.0.8 (2015/05/08)

Enhancements

* Add `filter_slack_quote`

# 0.0.7 (2015/04/02)

Fixes:

* filter_string_irc_slack: fix color code replacement

# 0.0.6 (2015/03/31)

Enhancements

* Add `filter_string_irc_slack`
* log.error InternalServerError

# 0.0.5 (2015/03/30)

Changes:

* Use Rack::Request to handle both `application/x-www-form-urlencoded` and `multipart/form-data`.

# 0.0.4 (2015/03/30)

Enhancements:

* Support `multipart/form-data` as ikachan does

# 0.0.3 (2015/03/25)

Changes:

* fluent-plugin-irc v0.0.7 has been released. Use it.

# 0.0.2 (2015/03/23)

Enhancements:

* Fallback to `#{nick}_` when `:err_nick_name_in_use` occurs

# 0.0.1 (2015/03/22)

Initial version



test.sh
#!/bin/sh
curl -d 'channel=#notify&message=*test* _daioikachan_ `www-form-urlencoded`' http://localhost:4979/notice
curl -F 'channel=#notify' -F 'message=*test* _daioikachan_ `multipart/form-data`' http://localhost:4979/privmsg


test/helper.rb
require 'test/unit'
require 'fileutils'
require 'fluent/log'
require 'fluent/test'

unless defined?(Test::Unit::AssertionFailedError)
  class Test::Unit::AssertionFailedError < StandardError
  end
end

def unused_port
  s = TCPServer.open(0)
  port = s.addr[1]
  s.close
  port
end

def ipv6_enabled?
  require 'socket'

  begin
    TCPServer.open("::1", 0)
    true
  rescue
    false
  end
end

$log = Fluent::Log.new(Fluent::Test::DummyLogDevice.new, Fluent::Log::LEVEL_WARN)


test/plugin/test_in_daioikachan.rb
require_relative '../helper'
require 'fluent/plugin/in_daioikachan'
require 'net/https'

class DaioikachanInputTest < Test::Unit::TestCase
  def post(path, params, header = {}, ssl = false)
    http = Net::HTTP.new("127.0.0.1", PORT)
    if ssl
      http.use_ssl = true
      http.verify_mode = OpenSSL::SSL::VERIFY_NONE
    end
    req = Net::HTTP::Post.new(path, header)
    if params.is_a?(String)
      req.body = params
    else
      req.set_form_data(params)
    end
    http.request(req)
  end

  def post_multipart(path, params, header = {}, ssl = false)
    http = Net::HTTP.new("127.0.0.1", PORT)
    if ssl
      http.use_ssl = true
      http.verify_mode = OpenSSL::SSL::VERIFY_NONE
    end
    req = Net::HTTP::Post.new(path, header)
    req.set_content_type("multipart/form-data; boundary=myboundary")

    body = ""
    params.each do |key, val|
      body.concat("--myboundary\r\n")
      body.concat("content-disposition: form-data; name=\"#{key}\";\r\n")
      body.concat("\r\n")
      body.concat("#{val}\r\n")
    end
    body.concat("--myboundary--\r\n")
    req.body = body

    http.request(req)
  end

  def setup
    Fluent::Test.setup
  end

  PORT = unused_port
  CONFIG = %[
    port #{PORT}
  ]

  def create_driver(conf = CONFIG)
    Fluent::Test::InputTestDriver.new(Fluent::DaioikachanInput).configure(conf, true)
  end

  def test_default_configure
    d = create_driver(%[])
    assert_equal 4979,      d.instance.port
    assert_equal '0.0.0.0', d.instance.bind
    assert_equal 0,         d.instance.min_threads
    assert_equal 4,         d.instance.max_threads
    assert_equal nil,       d.instance.backlog
  end

  def test_configure
    d = create_driver(%[
      port #{PORT}
      bind 127.0.0.1
      min_threads 0
      max_threads 4
      backlog 1024
    ])
    assert_equal PORT,        d.instance.port
    assert_equal '127.0.0.1', d.instance.bind
    assert_equal 0,           d.instance.min_threads
    assert_equal 4,           d.instance.max_threads
    assert_equal 1024,        d.instance.backlog
  end

  def test_notice
    d = create_driver

    time = Time.parse("2011-01-02 13:14:15 UTC").to_i
    Fluent::Engine.now = time

    d.expect_emit "notice.channel", time, {"command" => "notice", "channel" => "channel", "message" => "message"}

    d.run do
      d.expected_emits.each {|tag, time, record|
        res = post("/notice", {command: "notice", channel: "channel", message: "message"})
        assert_equal "200", res.code
      }
    end
  end

  def test_privmsg
    d = create_driver

    time = Time.parse("2011-01-02 13:14:15 UTC").to_i
    Fluent::Engine.now = time

    d.expect_emit "privmsg.channel", time, {"command" => "privmsg", "channel" => "channel", "message" => "message"}

    d.run do
      d.expected_emits.each {|tag, time, record|
        res = post("/privmsg", {command: "privmsg", channel: "channel", message: "message"})
        assert_equal "200", res.code
      }
    end
  end

  def test_join
    d = create_driver
    d.run do
      res = post("/join", {})
      assert_equal "200", res.code
    end
  end

  def test_leave
    d = create_driver
    d.run do
      res = post("/leave", {})
      assert_equal "200", res.code
    end
  end


  def test_multipart
    d = create_driver

    time = Time.parse("2011-01-02 13:14:15 UTC").to_i
    Fluent::Engine.now = time

    d.expect_emit "privmsg.channel", time, {"command" => "privmsg", "channel" => "channel", "message" => "message"}

    d.run do
      d.expected_emits.each {|tag, time, record|
        res = post_multipart("/privmsg", {command: "privmsg", channel: "channel", message: "message"})
        assert_equal "200", res.code
      }
    end
  end
end


lib/fluent/plugin/out_irc.rb
module Fluent
  class IRCOutput < Fluent::Output
    Fluent::Plugin.register_output('irc', self)

    include SetTimeKeyMixin
    include SetTagKeyMixin

    config_set_default :include_time_key, true
    config_set_default :include_tag_key, true

    config_param :host        , :string  , :default => 'localhost'
    config_param :port        , :integer , :default => 6667
    config_param :channel     , :string
    config_param :channel_keys, :default => nil do |val|
      val.split(',')
    end
    config_param :nick        , :string  , :default => 'fluentd'
    config_param :user        , :string  , :default => 'fluentd'
    config_param :real        , :string  , :default => 'fluentd'
    config_param :password    , :string  , :default => nil
    config_param :message     , :string
    config_param :out_keys do |val|
      val.split(',')
    end
    config_param :time_key    , :string  , :default => 'time'
    config_param :time_format , :string  , :default => '%Y/%m/%d %H:%M:%S'
    config_param :tag_key     , :string  , :default => 'tag'
    config_param :command     , :string  , :default => 'privmsg'
    config_param :command_keys, :default => nil do |val|
      val.split(',')
    end

    config_param :blocking_timeout, :time,    :default => 0.5
    config_param :max_send_queue,   :integer, :default => 100
    config_param :send_interval,    :time,    :default => 2

    COMMAND_MAP = {
      'priv_msg' => :priv_msg,
      'privmsg'  => :priv_msg,
      'notice'   => :notice,
    }

    # To support log_level option implemented by Fluentd v0.10.43
    unless method_defined?(:log)
      define_method("log") { $log }
    end

    attr_reader :conn # for test

    def initialize
      super
      require 'irc_parser'
    end

    def configure(conf)
      super

      begin
        @message % (['1'] * @out_keys.length)
      rescue ArgumentError
        raise Fluent::ConfigError, "string specifier '%s' and out_keys specification mismatch"
      end

      if @channel_keys
        begin
          @channel % (['1'] * @channel_keys.length)
        rescue ArgumentError
          raise Fluent::ConfigError, "string specifier '%s' and channel_keys specification mismatch"
        end
      end
      @channel = '#'+@channel

      if @command_keys
        begin
          @command % (['1'] * @command_keys.length)
        rescue ArgumentError
          raise Fluent::ConfigError, "string specifier '%s' and command_keys specification mismatch"
        end
      else
        unless @command = COMMAND_MAP[@command]
          raise Fluent::ConfigError, "command must be one of #{COMMAND_MAP.keys.join(', ')}"
        end
      end

      @send_queue = []
    end

    def start
      super

      begin
        @loop = Coolio::Loop.new
        @conn = create_connection
        @timer = TimerWatcher.new(@send_interval, true, log, &method(:on_timer))
        @loop.attach(@timer)
        @thread = Thread.new(&method(:run))
      rescue => e
        puts e
        raise Fluent::ConfigError, "failed to connect IRC server #{@host}:#{@port}"
      end
    end

    def shutdown
      super
      @loop.watchers.each { |w| w.detach }
      @loop.stop
      @conn.close
      @thread.join
    end

    def run
      @loop.run(@blocking_timeout)
    rescue => e
      log.error "unexpected error", :error => e, :error_class => e.class
      log.error_backtrace
    end

    def emit(tag, es, chain)
      chain.next

      if @conn.closed?
        log.warn "out_irc: connection is closed. try to reconnect"
        @conn = create_connection
      end

      es.each do |time,record|
        if @send_queue.size >= @max_send_queue
          log.warn "out_irc: send queue size exceeded max_send_queue(#{@max_send_queue}), discards"
          break
        end

        filter_record(tag, time, record)
        command, channel, message = build_command(record), build_channel(record), build_message(record)
        log.debug { "out_irc: push {command:\"#{command}\", channel:\"#{channel}\", message:\"#{message}\"}" }
        @send_queue.push([command, channel, message])
      end
    end

    def on_timer
      return if @send_queue.empty?
      command, channel, message = @send_queue.shift
      log.info { "out_irc: send {command:\"#{command}\", channel:\"#{channel}\", message:\"#{message}\"}" }
      @conn.send_message(command, channel, message)
    end

    private

    def create_connection
      conn = IRCConnection.connect(@host, @port)
      conn.log  = log
      conn.nick = @nick
      conn.user = @user
      conn.real = @real
      conn.password = @password
      conn.attach(@loop)
      conn
    end

    def build_message(record)
      values = fetch_keys(record, @out_keys)
      @message % values
    end

    def build_channel(record)
      return @channel unless @channel_keys

      values = fetch_keys(record, @channel_keys)
      @channel % values
    end

    def build_command(record)
      return @command unless @command_keys

      values = fetch_keys(record, @command_keys)
      unless command = COMMAND_MAP[@command % values]
        log.warn "out_irc: command is not one of #{COMMAND_MAP.keys.join(', ')}, use privmsg"
      end
      command || :priv_msg
    end

    def fetch_keys(record, keys)
      Array(keys).map do |key|
        begin
          record.fetch(key).to_s
        rescue KeyError
          log.warn "out_irc: the specified key '#{key}' not found in record. [#{record}]"
          ''
        end
      end
    end

    class TimerWatcher < Coolio::TimerWatcher
      def initialize(interval, repeat, log, &callback)
        @callback = callback
        @log = log
        super(interval, repeat)
      end

      def on_timer
        @callback.call
      rescue
        @log.error $!.to_s
        @log.error_backtrace
      end
    end

    class IRCConnection < Cool.io::TCPSocket
      attr_reader :joined # for test
      attr_accessor :log, :nick, :user, :real, :password

      def initialize(*args)
        super
        @joined = {}
      end

      def on_connect
        if @password
          IRCParser.message(:pass) do |m|
            m.password = @password
            write m
          end
        end
        IRCParser.message(:nick) do |m|
          m.nick   = @nick
          write m
        end
        IRCParser.message(:user) do |m|
          m.user = @user
          m.postfix = @real
          write m
        end
      end

      def on_read(data)
        err_nick_name_in_use = nil
        data.each_line do |line|
          begin
            msg = IRCParser.parse(line)
            log.debug { "out_irc: on_read :#{msg.class.to_sym}" }
            case msg.class.to_sym
            when :rpl_welcome
              log.info { "out_irc: welcome \"#{msg.nick}\" to \"#{msg.prefix}\"" }
            when :ping
              IRCParser.message(:pong) do |m|
                m.target = msg.target
                m.body = msg.body
                write m
              end
            when :join
              log.info { "out_irc: joined to #{msg.channels.join(', ')}" }
              msg.channels.each {|channel| @joined[channel] = true }
            when :err_nick_name_in_use
              log.warn "out_irc: nickname \"#{msg.error_nick}\" is already in use. use \"#{msg.error_nick}_\" instead."
              err_nick_name_in_use = true
              @nick = "#{msg.error_nick}_"
            when :error
              log.warn "out_irc: an error occured. \"#{msg.error_message}\""
            end
          rescue
            #TODO
          end
        end

        if err_nick_name_in_use
          self.on_connect
        end
      end

      def joined?(channel)
        @joined[channel]
      end

      def join(channel)
        IRCParser.message(:join) do |m|
          m.channels = channel
          write m
        end
        log.debug { "out_irc: join to #{channel}" }
      end

      def send_message(command, channel, message)
        join(channel) unless joined?(channel)
        IRCParser.message(command) do |m|
          m.target = channel
          m.body = message
          write m
        end
        channel # return channel for test
      end
    end
  end
end


lib/fluent/plugin/out_slack.rb
require_relative 'slack_client'

module Fluent
  class SlackOutput < Fluent::TimeSlicedOutput
    Fluent::Plugin.register_output('buffered_slack', self) # old version compatiblity
    Fluent::Plugin.register_output('slack', self)

    include SetTimeKeyMixin
    include SetTagKeyMixin

    config_set_default :include_time_key, true
    config_set_default :include_tag_key, true
   
    config_param :webhook_url,   :string, default: nil # incoming webhook
    config_param :token,         :string, default: nil # api token
    config_param :username,      :string, default: 'fluentd'
    config_param :color,         :string, default: 'good'
    config_param :icon_emoji,    :string, default: ':question:'

    config_param :channel,       :string
    config_param :channel_keys,  default: nil do |val|
      val.split(',')
    end
    config_param :title,         :string, default: nil
    config_param :title_keys,    default: nil do |val|
      val.split(',')
    end
    config_param :message,       :string, default: nil
    config_param :message_keys,  default: nil do |val|
      val.split(',')
    end

    # for test
    attr_reader :slack, :time_format

    def initialize
      super
      require 'uri'
    end

    def configure(conf)
      @time_format ||= conf['time_format'] ||= '%H:%M:%S' # old version compatiblity
 
      super

      @channel = URI.unescape(@channel) # old version compatibility
      @channel = '#' + @channel unless @channel.start_with?('#')

      if @webhook_url
        # following default values are for old version compatibility
        @title         ||= '%s'
        @title_keys    ||= %w[tag]
        @message       ||= '[%s] %s'
        @message_keys  ||= %w[time message]
        @slack = Fluent::SlackClient::IncomingWebhook.new(@webhook_url)
      else
        unless @token
          raise Fluent::ConfigError.new("`token` is required to call slack api")
        end
        @message      ||= '%s'
        @message_keys ||= %w[message]
        @slack = Fluent::SlackClient::WebApi.new
      end

      begin
        @message % (['1'] * @message_keys.length)
      rescue ArgumentError
        raise Fluent::ConfigError, "string specifier '%s' for `message`  and `message_keys` specification mismatch"
      end
      if @title and @title_keys
        begin
          @title % (['1'] * @title_keys.length)
        rescue ArgumentError
          raise Fluent::ConfigError, "string specifier '%s' for `title` and `title_keys` specification mismatch"
        end
      end
      if @channel_keys
        begin
          @channel % (['1'] * @channel_keys.length)
        rescue ArgumentError
          raise Fluent::ConfigError, "string specifier '%s' for `channel` and `channel_keys` specification mismatch"
        end
      end
    end

    def format(tag, time, record)
      [tag, time, record].to_msgpack
    end

    def write(chunk)
      begin
        payloads = build_payloads(chunk)
        payloads.each {|payload| @slack.post_message(payload) }
      rescue Net::OpenTimeout, Net::ReadTimeout => e
        log.warn "out_slack:", :error => e.to_s, :error_class => e.class.to_s
        raise e # let Fluentd retry
      rescue => e
        log.error "out_slack:", :error => e.to_s, :error_class => e.class.to_s
        log.warn_backtrace e.backtrace
        # discard. @todo: add more retriable errors
      end
    end

    private

    def build_payloads(chunk)
      if @title
        build_title_payloads(chunk)
      else
        build_plain_payloads(chunk)
      end
    end

    def common_payload
      return @common_payload if @common_payload
      @common_payload = {
        username:   @username,
        icon_emoji: @icon_emoji,
      }
      @common_payload[:token] = @token if @token
      @common_payload
    end

    Field = Struct.new("Field", :title, :value)

    def build_title_payloads(chunk)
      ch_fields = {}
      chunk.msgpack_each do |tag, time, record|
        channel = build_channel(record)
        per     = tag # title per tag
        ch_fields[channel]      ||= {}
        ch_fields[channel][per] ||= Field.new(build_title(record), '')
        ch_fields[channel][per].value << "#{build_message(record)}\n"
      end
      ch_fields.map do |channel, fields|
        {
          channel: channel,
          attachments: [{
            :color    => @color,
            :fallback => fields.values.map(&:title).join(' '), # fallback is the message shown on popup
            :fields   => fields.values.map(&:to_h)
          }],
        }.merge(common_payload)
      end
    end

    def build_plain_payloads(chunk)
      messages = {}
      chunk.msgpack_each do |tag, time, record|
        channel = build_channel(record)
        messages[channel] ||= ''
        messages[channel] << "#{build_message(record)}\n"
      end
      messages.map do |channel, text|
        {
          channel: channel,
          attachments: [{
            :color    => @color,
            :fallback => text,
            :text     => text,
          }],
        }.merge(common_payload)
      end
    end

    def build_message(record)
      values = fetch_keys(record, @message_keys)
      @message % values
    end

    def build_title(record)
      return @title unless @title_keys

      values = fetch_keys(record, @title_keys)
      @title % values
    end

    def build_channel(record)
      return @channel unless @channel_keys

      values = fetch_keys(record, @channel_keys)
      @channel % values
    end

    def fetch_keys(record, keys)
      Array(keys).map do |key|
        begin
          record.fetch(key).to_s
        rescue KeyError
          log.warn "out_slack: the specified key '#{key}' not found in record. [#{record}]"
          ''
        end
      end
    end
  end
end


lib/fluent/plugin/slack_client.rb
require 'uri'
require 'net/http'

module Fluent
  module SlackClient
    class Error < StandardError; end

    # This slack client only supports posting message
    class Base
      attr_accessor :log, :debug_dev

      DEFAULT_ENDPOINT = "https://slack.com/api/chat.postMessage"

      def initialize(endpoint = nil)
        # Configure Incoming Webhook endpoint instead of chat.postMessage API
        @endpoint = URI.parse(endpoint) if endpoint
      end

      def endpoint
        @endpoint ||= URI.parse(DEFAULT_ENDPOINT)
      end

      # Sends a message to a channel.
      #
      # @option params [channel] :channel
      #   Channel to send message to. Can be a public channel, private group or IM channel. Can be an encoded ID, or a name.
      # @option params [Object] :text
      #   Text of the message to send. See below for an explanation of formatting.
      # @option params [Object] :username
      #   Name of bot.
      # @option params [Object] :parse
      #   Change how messages are treated. See below.
      # @option params [Object] :link_names
      #   Find and link channel names and usernames.
      # @option params [Object] :attachments
      #   Structured message attachments.
      # @option params [Object] :unfurl_links
      #   Pass true to enable unfurling of primarily text-based content.
      # @option params [Object] :unfurl_media
      #   Pass false to disable unfurling of media content.
      # @option params [Object] :icon_url
      #   URL to an image to use as the icon for this message
      # @option params [Object] :icon_emoji
      #   emoji to use as the icon for this message. Overrides `icon_url`.
      # @see https://api.slack.com/methods/chat.postMessage
      # @see https://github.com/slackhq/slack-api-docs/blob/master/methods/chat.postMessage.md
      # @see https://github.com/slackhq/slack-api-docs/blob/master/methods/chat.postMessage.json
      def post_message(params = {})
        http = Net::HTTP.new(endpoint.host, endpoint.port)
        http.use_ssl = (endpoint.scheme == 'https')
        http.verify_mode = OpenSSL::SSL::VERIFY_NONE
        http.set_debug_output(debug_dev) if debug_dev

        req = Net::HTTP::Post.new(endpoint.path)
        req['Host'] = endpoint.host
        req['Accept'] = 'application/json; charset=utf-8'
        req['User-Agent'] = 'fluent-plugin-slack'
        req.body = encode_body(params)

        log.info { "out_slack: post #{params.dup.tap {|p| p[:token] = '[FILTERED]' if p[:token] }}" }
        res = http.request(req)
        response_check(res)
      end

      def encode_body(params)
        raise NotImplementedError
      end

      def response_check(res)
        if res.code != "200"
          raise Error, "Slack.com - #{res.code} - #{res.body}"
        end
      end
    end

    class IncomingWebhook < Base
      def encode_body(params = {})
        params.to_json
      end

      def response_check(res)
        super
        unless res.body == 'ok'
          raise Error, "Slack.com - #{res.code} - #{res.body}"
        end
      end
    end

    class WebApi < Base
      def encode_body(params = {})
        body = params.dup
        if params[:attachments]
          body[:attachments] = params[:attachments].to_json
        end
        URI.encode_www_form(body)
      end

      def response_check(res)
        super
        params = JSON.parse(res.body)
        unless params['ok']
          raise Error, "Slack.com - #{res.code} - #{res.body}"
        end
      end
    end
  end
end


lib/fluent/plugin/filter_string_irc_slack.rb
require 'string/scrub' if RUBY_VERSION.to_f < 2.1

module Fluent
  class StringIrcSlackFilter < Filter
    Plugin.register_filter('string_irc_slack', self)

    COLOR_CODE = "\x03" # \u0003
    BOLD       = "\x02" # \u0002
    UNDERLINE  = "\x1f" # \u001F
    INVERSE    = "\x16" # \u0016
    CLEAR      = "\x0f" # \u000F

    def configure(conf)
      @start_code = Regexp.new("(#{COLOR_CODE}[0-9]?[0-9](,[0-9]?[0-9])?|#{BOLD}|#{UNDERLINE}|#{INVERSE})+")
      @stop_code  = Regexp.new(CLEAR)
      super
    end

    def start
      super
    end

    def shutdown
      super
    end

    def filter(tag, time, record)
      if message = record['message']
        filtered_message = with_scrub(message) {|str| str.gsub(@start_code, '`').gsub(@stop_code, '`') }
        record = record.dup.tap {|r| r['message'] = filtered_message }
      end
      record
    end

    def with_scrub(string)
      begin
        return yield(string)
      rescue ArgumentError => e
        raise e unless e.message.index("invalid byte sequence in") == 0
        log.info "filter_string_irc_slack: invalid byte sequence is replaced in #{string}"
        string.scrub!('?')
        retry
      end
    end
  end
end


lib/fluent/plugin/in_daioikachan.rb
module Fluent
  class DaioikachanInput < Input
    Plugin.register_input('daioikachan', self)

    def initialize
      require 'puma'
      require 'uri'
      require 'rack'
      super
    end

    config_param :port, :integer, :default => 4979
    config_param :bind, :string, :default => '0.0.0.0'
    config_param :min_threads, :integer, :default => 0
    config_param :max_threads, :integer, :default => 4
    config_param :backlog, :integer, :default => nil

    def configure(conf)
      super
    end

    def start
      super

      # Refer puma's Runner and Rack handler for puma server setup
      @server = ::Puma::Server.new(method(:on_request))
      @server.min_threads = @min_threads
      @server.max_threads = @max_threads
      @server.leak_stack_on_error = false
      setup_http

      @app = App.new(self)

      @thread = Thread.new(&method(:run))
    end

    def shutdown
      @server.stop(true)
      @thread.join
    end

    def run
      @server.run(false)
    rescue => e
      log.error "unexpected error", :error => e.to_s
      log.error_backtrace e.backtrace
    end

    def on_request(env)
      log.debug { "in_daioikachan: #{env.to_s}" }
      @app.run(env)
    end

    def setup_http
      log.info "listening http on #{@bind}:#{@port}"

      opts = [@bind, @port, true]
      opts << @backlog if @backlog
      @server.add_tcp_listener(*opts)
    end

    class App
      class BadRequest < StandardError; end

      attr_reader :router, :log

      def initialize(plugin)
        @router = plugin.router
        @log    = plugin.log
      end

      def run(env)
        req = Rack::Request.new(env)
        begin
          if req.post?
            params = req.params
            case req.path
            when '/notice'
              notice(params)
            when '/privmsg'
              privmsg(params)
            when '/join'
              return ok
            when '/leave'
              return ok
            else
              return not_found
            end
          else
            return not_found
          end
        rescue BadRequest => e
          bad_request(e.message)
        rescue => e
          log.error "out_slack:", :error => e.to_s, :error_class => e.class.to_s
          log.warn_backtrace e.backtrace
          internal_server_error
        else
          ok
        end
      end

      def notice(params)
        channel, message = build_channel(params), build_message(params)
        tag    = "notice.#{channel}"
        record = params.merge('command' => 'notice', 'channel' => channel, 'message' => message)
        router.emit(tag, Fluent::Engine.now, record)
      end

      def privmsg(params)
        channel, message = build_channel(params), build_message(params)
        tag    = "privmsg.#{channel}"
        record = params.merge('command' => 'privmsg', 'channel' => channel, 'message' => message)
        router.emit(tag, Fluent::Engine.now, record)
      end

      private

      def build_channel(params)
        unless channel = params.delete('channel')
          raise BadRequest.new('`channel` parameter is mandatory')
        end
        if channel.start_with?('#')
          channel[1..-1] # remove starting #
        else
          channel
        end
      end

      def build_message(params)
        unless message = params.delete('message')
          raise BadRequest.new('`message` parameter is mandatory')
        end
        message # should I truncate message to max_length?
      end

      def ok(msg = nil)
        [200, {'Content-type'=>'text/plain'}, ["OK\n#{msg}"]]
      end

      def bad_request(msg = nil)
        [400, {'Content-type'=>'text/plain'}, ["Bad Request\n#{msg}"]]
      end

      def not_found(msg = nil)
        [404, {'Content-type'=>'text/plain'}, ["Not Found\n#{msg}"]]
      end

      def internal_server_error(msg = nil)
        [500, {'Content-type'=>'text/plain'}, ["Internal Server Error\n#{msg}"]]
      end
    end
  end
end
